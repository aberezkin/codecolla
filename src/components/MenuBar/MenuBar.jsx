import React, {Component} from 'react';

export const MENU_BAR_CLASSNAME = 'MenuBar';

class MenuBar extends Component {
    constructor(props) {
        super(props);
        this.renderMenuItem = this.renderMenuItem.bind(this);
        this.bindSetInactiveHandler = this.bindSetInactiveHandler.bind(this);
        this.unbindSetInactiveHandler = this.unbindSetInactiveHandler.bind(this);
        this.onClick = this.onClick.bind(this);
        this.handleDocumentClick = this.handleDocumentClick.bind(this);
        this.state = {isActive: false};
    }
    componentDidUpdate(prevProps, prevState) {
        if (this.state.isActive && !prevState.isActive) {
            this.bindSetInactiveHandler()
        }

        else if (prevState.isActive && !this.state.isActive) {
            this.unbindSetInactiveHandler()
        }
    }

    renderMenuItem(child) {
        return React.cloneElement(child, {
            isMenuBarActive: this.state.isActive, //is MenuBar 'opened'
            isTopLevel: true,                     //is element on top
            onSelect: this.props.onSelect,        //callback forall commands
        });
    }

    bindSetInactiveHandler() {
        document.addEventListener('click', this.handleDocumentClick, false);
    }

    unbindSetInactiveHandler() {
        document.removeEventListener('click', this.handleDocumentClick);
    }

    handleDocumentClick(event) {
        this.setState({isActive: false});
    }

    onClick(event) {
        this.setState({isActive: !this.state.isActive});
    }

    render() {
        return (
            <ul className={`${MENU_BAR_CLASSNAME}`} onClick={this.onClick}>
                {React.Children.map(this.props.children, this.renderMenuItem)}
            </ul>
        )
    }
}

MenuBar.defaultProps = {
    onSelect: (command) => {
        console.log(command);
    },
};

export default MenuBar;