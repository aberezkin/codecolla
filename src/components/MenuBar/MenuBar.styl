.MenuBar
  background inherit
  color inherit
  width 100%
  height 20px
  position relative
  border-bottom 1px solid black