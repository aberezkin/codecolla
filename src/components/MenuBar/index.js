export { default as MenuBar } from './MenuBar';
export { default as SubMenu } from './SubMenu/SubMenu';
export { default as Checkable } from './Checkable/Checkable';
export { default as MenuItem } from './MenuItem/MenuItem';
export { default as Separator } from './Separator/Separator';
