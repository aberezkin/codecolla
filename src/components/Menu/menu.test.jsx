import React from 'react';
import { MenuBar, MenuItem, SubMenu, Checkable } from '../MenuBar';

const setup = (propChildren) => {
    const props = {
        children: [<MenuItem key={0} title="Test" command="test" />],
        onSelect: jest.fn(),
        ...propChildren,
    };

    const wrapper = mount((
        <MenuBar onSelect={props.onSelect}>
            {props.children}
        </MenuBar>
    ));

    return {
        onSelect: props.onSelect,
        wrapper,
    };
};

describe('menu component', () => {
    it('render MenuBar', () => {
        const { wrapper } = setup();

        expect(wrapper).toMatchSnapshot();
    });

    it('check rendering of MenuItem', () => {
        const { wrapper } = setup();

        expect(wrapper.contains([<MenuItem title="Test" command="test" />])).toBeTruthy();
    });

    it('should call onSelect when click MenuItem "test"', () => {
        const { onSelect, wrapper } = setup();

        wrapper.find('.MenuItem .title').simulate('click');
        expect(onSelect).toHaveBeenCalled();
    });

    it('should not call onSelect when click MenuItem without command', () => {
        const children = [<MenuItem title="Test" key={0} />];

        const { onSelect, wrapper } = setup({ children });

        wrapper.find('.MenuItem .title').simulate('click');
        expect(onSelect).not.toHaveBeenCalled();
    });

    it('should not call onSelect when click Checkable without command', () => {
        const children = [
            <MenuItem title="Test" key={0}>
                <SubMenu>
                    <Checkable title="Checkable-test" />
                </SubMenu>
            </MenuItem>,
        ];

        const { onSelect, wrapper } = setup({ children });

        wrapper.find('.MenuItem .title').simulate('click');
        wrapper.find('.Checkable .title').simulate('click');
        expect(onSelect).not.toHaveBeenCalled();
    });

    it('should call onSelect when click Checkable "test"', () => {
        const children = [
            <MenuItem title="Test" key={0}>
                <SubMenu>
                    <Checkable title="Checkable-test" command="check" />
                </SubMenu>
            </MenuItem>,
        ];

        const { onSelect, wrapper } = setup({ children });

        wrapper.find('.MenuItem .title').simulate('click');
        expect(onSelect).not.toHaveBeenCalled();

        wrapper.find('.Checkable .title').simulate('click');
        expect(onSelect).toHaveBeenCalled();
    });
});
