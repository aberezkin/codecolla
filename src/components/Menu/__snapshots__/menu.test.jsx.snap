// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`menu component render MenuBar 1`] = `
<MenuBar
  onSelect={[Function]}
>
  <ul
    className="MenuBar"
    onClick={[Function]}
  >
    <MenuItem
      command="test"
      isMenuBarActive={false}
      isTopLevel={true}
      key=".$0"
      onSelect={[Function]}
      title="Test"
    >
      <li
        className="MenuItem"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseOut={[Function]}
        onMouseOver={[Function]}
      >
        <span
          className="title"
          onClick={[Function]}
          role="button"
          style={Object {}}
          tabIndex="0"
        >
          Test
        </span>
      </li>
    </MenuItem>
  </ul>
</MenuBar>
`;
